packages/db/src/indexes/auto-index.ts

```diff
@@ -0,0 +1,108 @@
+import { OrderedIndex } from "./ordered-index"
+import type { BasicExpression } from "../query/ir"
+import type { CollectionImpl } from "../collection"
+
+export interface AutoIndexConfig {
+  autoIndex?: `off` | `eager`
+}
+
+/**
+ * Analyzes a where expression and creates indexes for all simple operations on single fields
+ */
+export function ensureIndexForExpression<
+  T extends Record<string, any>,
+  TKey extends string | number,
+>(
+  expression: BasicExpression,
+  collection: CollectionImpl<T, TKey, any, any, any>
+): void {
+  // Only proceed if auto-indexing is enabled
+  if (collection.config.autoIndex !== `eager`) {
+    return
+  }
+
+  // Don't auto-index during sync operations
+  if (
+    collection.status === `loading` ||
+    collection.status === `initialCommit`
+  ) {
+    return
+  }
+
+  // Extract all indexable expressions and create indexes for them
+  const indexableExpressions = extractIndexableExpressions(expression)
+
+  for (const { fieldName, fieldPath } of indexableExpressions) {
+    // Check if we already have an index for this field
+    const existingIndex = Array.from(collection.indexes.values()).find(
```

Comment:
Potentially expensive operation here as we have to loop over all indexes to find the one that matches `fieldPath`. Could the index not keep a map from `fieldPath` to the index such that this becomes an O(1) lookup?